File: 155-min-stack/155-min-stack.py
class MinStack:

    def __init__(self):
        self.data = []
        self.min = []
        self.curr_min = float('inf')

    def push(self, val: int) -> None:
        self.data.append(val)
        self.curr_min = min(self.curr_min, val)
        self.min.append(self.curr_min)
        

    def pop(self) -> None:
        self.data.pop()
        self.min.pop()
        self.curr_min = self.min[-1] if self.min else sys.maxsize

    def top(self) -> int:
        return self.data[-1]

    def getMin(self) -> int:
        return self.min[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 155-min-stack/155-min-stack.java
class MinStack {
    Deque<Integer> stack;
    Deque<Integer> mins;
    int min;
    public MinStack() {
        stack = new ArrayDeque<>();
        mins = new ArrayDeque<>();
        min = Integer.MAX_VALUE;
    }
    
    public void push(int val) {
        stack.push(val);
        if(val < min)
            min = val;
        mins.push(min);
    }
    
    public void pop() {
        stack.pop();
        mins.pop();
        min = mins.isEmpty() ? Integer.MAX_VALUE : mins.peek();
    }
    
    public int top() {
        return stack.peek();
    }
    
    public int getMin() {
        return min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */